velox/functions/prestosql/types/JsonType.cpp

```diff
@@ -62,57 +66,415 @@ void generateJsonTyped(
+    exec::EvalCtx* context,
+    const SelectivityVector& rows,
+    FlatVector<StringView>& flatResult) {
+  if constexpr (kind == TypeKind::ARRAY) {
+    castToJsonFromArray(input, context, rows, flatResult);
+  } else if constexpr (kind == TypeKind::MAP) {
+    castToJsonFromMap(input, context, rows, flatResult);
+  } else if constexpr (kind == TypeKind::ROW) {
+    castToJsonFromRow(input, context, rows, flatResult);
+  } else {
+    VELOX_NYI(
+        "Casting from {} to JSON is not supported yet.",
+        input.type()->toString());
+  }
+}
+
+// Estimates the total length of all Json strings for a vector of arrays or maps
+// from the Json strings of their elements. The elementsStrings vector is
+// wrapped in elementsIndices. This estimate is an upperbound of the actual
+// total length.
+FOLLY_ALWAYS_INLINE size_t estimateJsonStringSizeForArraysAndMaps(
+    const FlatVector<StringView>& elementsStrings,
+    const vector_size_t* elementsIndices,
+    const vector_size_t* topLevelOffsets,
+    const vector_size_t* topLevelSizes,
+    const SelectivityVector& topLevelRows) {
+  size_t sum = 0;
+
+  topLevelRows.applyToSelected([&](auto row) {
+    auto offset = topLevelOffsets[row];
+    auto size = topLevelSizes[row];
+    for (auto i = offset; i < offset + size; ++i) {
+      sum += elementsStrings.valueAt(elementsIndices[i]).size();
+    }
+
+    // Extra length for commas, semicolons, and brackets to be added in Json
+    // strings.
+    sum += size > 0 ? size + 1 : 2;
+  });
+
+  return sum;
+}
+
+// Translates the selected rows of vector into the corresponding rows of the
+// base of vector.
+void translateRowsToBase(
+    const DecodedVector& vector,
+    const SelectivityVector& rows,
+    SelectivityVector& baseRows) {
+  baseRows.clearAll();
+  rows.applyToSelected(
+      [&](auto row) { baseRows.setValid(vector.index(row), true); });
+  baseRows.updateBounds();
+}
+
+void castToJsonFromArray(
+    const BaseVector& input,
+    exec::EvalCtx* context,
+    const SelectivityVector& rows,
+    FlatVector<StringView>& flatResult) {
+  // input is guaranteed to be in flat encoding when passed in.
+  auto inputArray = input.as<ArrayVector>();
+
+  auto elements = inputArray->elements();
+  auto elementsRows =
+      functions::toElementRows(elements->size(), rows, inputArray);
+  exec::LocalDecodedVector elementsDecoded(context, *elements, elementsRows);
+
+  exec::LocalSelectivityVector elementsBaseRows(
+      context->execCtx(), elementsDecoded->base()->size());
+  translateRowsToBase(*elementsDecoded, elementsRows, *elementsBaseRows);
+
+  VectorPtr elementsResult;
+  BaseVector::ensureWritable(
+      *elementsBaseRows, JSON(), context->pool(), &elementsResult);
+  auto elementsStringResult = elementsResult->as<FlatVector<StringView>>();
+
+  VELOX_DYNAMIC_TYPE_DISPATCH(
+      castToJson,
+      elements->typeKind(),
+      *elementsDecoded->base(),
+      context,
+      *elementsBaseRows,
+      *elementsStringResult);
```

Comment:
Not required, but this sequence of operations is repeated 4x and could probably be factored out IMO. You might want to make it a struct so you can also have state. Something like:

```
struct AsJson {
  AsJson(eval::EvalCtx* context, VectorPtr input, SelectivityVector& rows) : decoded_(context, *input, rows) {
    exec::LocalSelectivityVector baseRows(context->execCtx(), decoded_->base()->size());
    translateRowsToBase(*decoded, rows, *baseRows);
    VectorPtr json;
    BaseVector::ensureWritable(*baseRows, JSON(), context->pool(), &json);
    this->json = elementsResult->as<FlatVector<StringView>>();
    VELOX_DYNAMIC_TYPE_DISPATCH(castToJson, input->typeKind(), *decoded_->base(), context, *baseRows, *this->json);
  }

  StringView at(vector_size_t i) { return json->valueAt(decoded->index(i)); }
  bool nullAt(vector_size_t i) { return decoded->isNullAt(i); }

  exec::LocalDecodedVector decoded;
  FlatVector<StringView> json;
};
```

You might also inline `translateRowsToBase`, since it would only be called from here. You'd use this struct like this:
```
  AsJson elements(context, *elementsRaw, elementRows);
  auto elementsStringSize = estimateJsonStringSizeForArraysAndMaps(*elements.json, elements.decoded->indices(), ...);
  ...
  proxy.append(elements.at(i));
```

Reply:
This is a good suggestion! I've revised the code accordingly. Thank you!